src/lib.rs

```diff
@@ -409,7 +409,7 @@ mod instance;
 pub mod marker;
 pub mod marshal;
 #[macro_use]
-pub mod once_cell;
+pub mod sync;
```

Comment:
Can we add a deprecation for this? This will be much nicer than a compile error.

Reply:
Do you think a `pub use sync as once_cell;` would suffice?

Reply:
(With the deprecation of course, I just meant instead of keeping both modules around and having the re-exports in there.)

Reply:
I'm not sure. Is there actually a way to deprecate this?

Reply:
Hhhmm, it seems to work neither way, i.e. both marking the `pub use sync as once_cell;` and the `pub use crate::sync::GILOnceCell` inside an otherwise empty `mod once_cell;` does not produce a deprecation warning for usages of `pyo3::once_cell:GILOnceCell`.

Does anybody have an idea of how to keep the old path alive but deprecated without copy & pasting the code?

Reply:
Does having `#[deprecated] mod once_cell { pub use crate::sync::GILOnceCell }` work?

If not, I suggest keep both around for backwards compatibility, there's a rustc issue _somewhere_ on deprecating exports which I've been tempted to pick up sometime so that we can clean up some of our old module hierarchy :smile:

Reply:
> Does having #[deprecated] mod once_cell { pub use crate::sync::GILOnceCell } work?

No, it did not seem to work. I suspect because re-exports are not "items" and only those can be deprecated?

> If not, I suggest keep both around for backwards compatibility

This is a heavy burden though. What if make a change to one copy but forget about the other?

Reply:
I mean just have `pub mod once_cell { pub use crate::sync::GILOnceCell }` as an item in `lib.rs`, with a "TODO deprecate this" just above?

Reply:
> I mean just have pub mod once_cell { pub use crate::sync::GILOnceCell } as an item in lib.rs, with a "TODO deprecate this" just above?

This I can certainly do.

Reply:
I think the relevant issue is https://github.com/rust-lang/rust/issues/30827 which I linked in the FIXME comment.